### ISSUE TITLE
Empty SRC field in a violation statement

### ISSUE THREAD

**Issue by stefjoosten** [AmpersandTarski/Ampersand#1193]
# Situation
I have a population in a spreadsheet that contains rows of type `Beheerseenheid`.
One of the columns, `voorzieningtype` contains the type of beheerseenheid.
I want to use that type to populate my script:
```Ampersand
RELATION voorzieningtype [Beheerseenheid*Beheerseenheidtype] [UNI]

CLASSIFY Applicatie, Kaartlaag, Koppeling, Netwerk, Randapparatuur, Systeem, Tool, Videobron ISA Beheerseenheid
RULE "Typering van Applicaties" :
  voorzieningtype;"Applicatie" |- V[Applicatie*Beheerseenheidtype]
VIOLATION (TXT "{EX} InsPair;voorzieningtype;Applicatie;", SRC I, TXT ";Beheerseenheidtype;", TGT I)
```
# What I expected
I expected this to work
# What happened
I got 342 runtime errors, which is exactly the number of `Applicatie`s in my population.
<img width="638" alt="afbeelding" src="https://user-images.githubusercontent.com/275768/126027423-0b845348-6acd-4fde-b831-d3c6c1b97a29.png">

# Why is this a bug?
Judging by the error message there is an error in the violation text. The `SRC I` part should be an atom from `voorzieningtype`, but it is empty.

# Version of Ampersand that was used
I used Ampersand-v4.1.5

**Comment by hanjoosten** [comment]
Just for the record, this issue is related with #1192.

**Comment by RieksJ** [comment]
This is known for a long time. This is why the ExecEngine function [SetConcept](https://github.com/AmpersandTarski/prototype/blob/development/bootstrap/files/execEngineFunctions.php#L301-L327) was created.

**Comment by Michiel-s** [comment]
> In the Ampersand semantics, I[Application*Application] is implicitly defined

Ok. It is defined indeed, but it's not a Relation right? It is an expression. You cannot define/add/remove population from I if I am correct. Is it allowed to specify in a ADL script `POPULATION I[Concept] CONTAINS ..`??

I think indeed the SetConcept function is of use for you here.

**Comment by Michiel-s** [comment]
@stefjoosten Could you close this issue if you agree?

**Comment by RieksJ** [comment]
Would this also somehow relate to #834?

**Comment by stefjoosten** [comment]
> > In the Ampersand semantics, I[Application*Application] is implicitly defined
> 
> Ok. It is defined indeed, but it's not a Relation right? It is an expression. You cannot define/add/remove population from I if I am correct. Is it allowed to specify in a ADL script `POPULATION I[Concept] CONTAINS ..`??
> 
> I think indeed the SetConcept function is of use for you here.

There is no reason to forbid inserting and deleting atoms from `I`. Hence it should be allowed. Inserting has exactly the semantics of the function SetConcept and deleting an atom has exactly the semantics of DelAtom.

**Comment by hanjoosten** [comment]
> 
> Is it allowed to specify in a ADL script `POPULATION I[Concept] CONTAINS ..`??
> 


Sure you can populate a concept. See [this documentation](https://ampersandtarski.gitbook.io/documentation/the-language-ampersand/the-population-statement) for the details.

**Comment by stefjoosten** [comment]
> Would this also somehow relate to #834?

No, that is a very different issue.

**Comment by Michiel-s** [comment]
> > Is it allowed to specify in a ADL script `POPULATION I[Concept] CONTAINS ..`??
> 
> Sure you can populate a concept. See [this documentation](https://ampersandtarski.gitbook.io/documentation/the-language-ampersand/the-population-statement) for the details.

Please pay attention to the POPULATION statement I stated... This is not to populate the concept, but the RELATION I[Concept*Concept]. Syntax for that would be: `POPULATION I[Person*Person] CONTAINS [("A", "A")]`

I just checked this, and it is not allowed by the compiler, as I expected.

So @stefjoosten states that `There is no reason to forbid inserting and deleting atoms from I`, but as the compiler doesn't allow it, let's not implement this in the prototype framework. That would require the compiler to first provide the "relation" defintions for `I` anyway.

Some more thoughts on this: how about inserting pairs in `V[A*B]`, what would that mean semantically? That means adding the src and tgt atoms in their concept sets, right? I don't think we should go this way.

Btw, this type of issues take up our valuable time. I think there are more important and urgent issues. So let's close and forget about this.

**Comment by hanjoosten** [comment]
You are right. An expression can never be populated with the `POPULATION` statement. The effect you describe would be denoted as `POPULATION Person CONTAINS ["A"]`
